autopilot/hostinfo.go

```diff
@@ -65,66 +65,72 @@ func (c *contractor) HostInfo(ctx context.Context, hostKey types.PublicKey) (api
-		return nil, fmt.Errorf("failed to fetch recommended fee from bus: %w", err)
+	// declare helper to decide whether to keep a host.
+	if usabilityMode != api.UsabilityFilterModeUsable && usabilityMode != api.UsabilityFilterModeUnusable && usabilityMode != api.UsabilityFilterModeAll {
+		return nil, fmt.Errorf("invalid usability mode: %v, options are usable and unusable", usabilityMode)
+	}
+	keep := func(usable bool) bool {
+		switch usabilityMode {
+		case api.UsabilityFilterModeUsable:
+			return usable // keep usable
+		case api.UsabilityFilterModeUnusable:
+			return !usable // keep unusable
+		case api.UsabilityFilterModeAll:
+			return true // keep all
+		default:
+			panic("unreachable")
+		}
 	}
 
-	f := newIPFilter(c.logger)
-	gc := worker.NewGougingChecker(gs, rs, cs, fee, cfg.Contracts.Period, cfg.Contracts.RenewWindow)
+	var hostInfos []api.HostHandlerGET
+	wanted := limit
+	for {
+		// fetch up to 'limit' hosts.
+		hosts, err := c.ap.bus.SearchHosts(ctx, filterMode, addressContains, keyIn, offset, limit)
+		if err != nil {
+			return nil, err
+		}
+		offset += len(hosts)
 
-	c.mu.Lock()
-	storedData := make(map[types.PublicKey]uint64)
-	for _, host := range hosts {
-		storedData[host.PublicKey] = c.cachedDataStored[host.PublicKey]
-	}
-	minScore := c.cachedMinScore
-	c.mu.Unlock()
+		// if there are no more hosts, we're done.
+		if len(hosts) == 0 {
+			return hostInfos, nil // no more hosts
+		}
 
-	var hostInfos []api.HostHandlerGET
-	for _, host := range hosts {
-		// ignore the pricetable's HostBlockHeight by setting it to our own blockheight
-		host.PriceTable.HostBlockHeight = cs.BlockHeight
+		// decide how many of the returned hosts to keep.
+		var keptHosts int
+		for _, host := range hosts {
+			hi, cached := hostInfo[host.PublicKey]
+			if !cached {
+				continue
+			}
+			if !keep(hi.Usable) {
+				continue
+			}
+			hostInfos = append(hostInfos, api.HostHandlerGET{
+				Host: host,
 
-		isUsable, unusableResult := isUsableHost(cfg, rs, gc, f, host, minScore, storedData[host.PublicKey])
-		hostInfos = append(hostInfos, api.HostHandlerGET{
-			Host: host,
+				Gouging:          hi.UnusableResult.gougingBreakdown.Gouging(),
+				GougingBreakdown: hi.UnusableResult.gougingBreakdown,
+				Score:            hi.UnusableResult.scoreBreakdown.Score(),
+				ScoreBreakdown:   hi.UnusableResult.scoreBreakdown,
+				Usable:           hi.Usable,
+				UnusableReasons:  hi.UnusableResult.reasons(),
+			})
+			if wanted > 0 && len(hostInfos) == wanted {
+				return hostInfos, nil // we're done.
+			}
+			keptHosts++
+		}
 
-			Gouging:          unusableResult.gougingBreakdown.Gouging(),
-			GougingBreakdown: unusableResult.gougingBreakdown,
-			Score:            unusableResult.scoreBreakdown.Score(),
-			ScoreBreakdown:   unusableResult.scoreBreakdown,
-			Usable:           isUsable,
-			UnusableReasons:  unusableResult.reasons(),
-		})
+		// if no hosts were kept from this batch, double the limit.
+		if keptHosts == 0 {
+			limit *= 2
```

Comment:
is it possible for limit to be -1? potentially breaking the behaviour of that sentinel value